native/dsyev.go

```diff
@@ -0,0 +1,116 @@
+// Copyright ©2016 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package native
+
+import (
+	"math"
+
+	"github.com/gonum/blas"
+	"github.com/gonum/blas/blas64"
+	"github.com/gonum/lapack"
+)
+
+// Dsyev computes all eigenvalues and, optionally, the eigenvectors of a real
+// symmetric matrix A.
+//
+// w contains the eigenvalues in ascending order upon return. w must have length
+// at least n, and Dsyev will panic otherwise.
+//
+// On entry, a contains the elements of the symmetric matrix A in the triangular
+// portion specified by uplo. If jobz == lapack.EigComp A contains the
+// orthonormal eigenvectors of A on exit, otherwise on exit the specified
+// triangular region is overwritten.
+//
+// Work is temporary storage, and lwork specifies the usable memory length. At minimum,
+// lwork >= 3*n-1, and Dsyev will panic otherwise. The amount of blocking is
+// limited by the usable length.
+// If lwork == -1, instead of computing Dsyev the optimal work length is stored
+// into work[0].
+func (impl Implementation) Dsyev(jobz lapack.EigComp, uplo blas.Uplo, n int, a []float64, lda int, w, work []float64, lwork int) (ok bool) {
+	checkMatrix(n, n, a, lda)
+	upper := uplo == blas.Upper
+	wantz := jobz == lapack.EigDecomp
+	var opts string
+	if upper {
+		opts = "U"
+	} else {
+		opts = "L"
+	}
+	nb := impl.Ilaenv(1, "DSYTRD", opts, n, -1, -1, -1)
+	lworkopt := max(1, (nb+2)*n)
+	work[0] = float64(lworkopt)
+	if lwork == -1 {
+		return
+	}
+	if len(work) < lwork {
+		panic(badWork)
+	}
+	if lwork < 3*n-1 {
+		panic(badWork)
+	}
+	if n == 0 {
+		return true
+	}
+	if n == 1 {
+		w[0] = a[0]
+		work[0] = 2
+		if wantz {
+			a[0] = 1
+			return true
+		}
+	}
+	safmin := dlamchS
+	eps := dlamchE
+	smlnum := safmin / eps
+	bignum := 1 / smlnum
+	rmin := math.Sqrt(smlnum)
+	rmax := math.Sqrt(bignum)
+
+	// Scale matrix to allowable range, if necessary.
+	anrm := impl.Dlansy(lapack.MaxAbs, uplo, n, a, lda, work)
+	type scaletype int
+	const (
+		none scaletype = iota
+		down
+		up
+	)
+	var sigma float64
+	iscale := none
+	if anrm > 0 && anrm < rmin {
+		iscale = down
+		sigma = rmin / anrm
+	} else if anrm > rmax {
+		iscale = up
```

Comment:
iscale can only be 1 or 0 in this code. Changed to a boolean with a better name.